### PR TITLE
Rewrite data management user guide a bit

### PR DESCRIPTION
**Pull request by benclifford**
putting this up for people to look at, but I want to work on it more

### CODE REVIEW

**Comment by danielskatz** [comment]
Is this still active?  And does it make sense to look at now?

**Comment by benclifford** [comment]
@danielskatz I just re-read this text, and I'm happy for it to be merged into master.

File: docs/userguide/data.rst [hunk 1]
```diff
@@ -3,7 +3,9 @@
 Data management
 ===============
 
-Parsl enables implementation of dataflow patterns in which data passed among apps manage the flow of execution. Dataflow programming models are popular as they can cleanly express, via implicit concurrency, opportunities for parallel execution.
+Parsl supports staging of file data between execution locations and at-rest
+locations. This supports two pieces of functionality: execution location
```
Comment (danielskatz) [inline docs/userguide/data.rst:7]: this sounds a little like the staging is one-way, and not the way I expected - maybe say more explicitly that staging is from at-rest to execution for inputs and back for outputs

File: docs/userguide/data.rst [hunk 2]
```diff
@@ -20,46 +22,75 @@ To this end, Parsl:
...
 
-The file may then be passed as input or output to an app. The following example executes the ``cat`` command on a local file:
 
-.. code-block:: python
+Staging providers
+-----------------
+
+Each executor can be configured with a list of
+:py:class:`~parsl.data_provider.staging.Staging` instances
+which will be used to stage files in and out of execution
```
Comment (danielskatz) [inline docs/userguide/data.rst:67]: for me, which should be that

File: docs/userguide/data.rst [hunk 3]
```diff
@@ -20,46 +22,75 @@ To this end, Parsl:
...
-    # create a test file
-    open('/tmp/test.txt', 'w').write('Hello\n')
+NoOpFileStaging for Local/Shared File Systems
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-    # create the Parsl file
-    parsl_file = File('file:///tmp/test.txt')
+The NoOpFileStaging provider assumes that files specified either
+with a path or with the ``file`` URL scheme are available both
+on the submit and execution side - for example, when there is a
```
Comment (danielskatz) [inline docs/userguide/data.rst:81]: maybe "this occurs, for example, when there is a"?

File: docs/userguide/data.rst [hunk 4]
```diff
@@ -87,12 +118,29 @@ The following example illustrates how the remote file is implicitly downloaded f
     f = convert(inputs=[inp], outputs=[out])
     f.result()
 
+FTP, HTTP, HTTPS: in-task staging
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+These staging providers are intended for use on executors which do not have
```
Comment (danielskatz) [inline docs/userguide/data.rst:124]: which should be that

File: docs/userguide/futures.rst [hunk 5]
```diff
@@ -105,11 +105,18 @@ That is, an app that is dependent upon the successful completion of another app
...
+While an AppFuture represents the execution of an asynchronous app, a DataFuture represents a file that an app produces.
 Parsl's dataflow model requires such a construct so that it can determine when other apps that are to consume a file produced by the app can start execution. 
-When calling an app that produces files as outputs, Parsl requires that a list of output files be specified via the ``outputs`` keyword argument. A DataFuture is returned for each file by the app when it executes. 
-As the app executes, Parsl monitors each file to 1) ensure it is created, and 2) pass it to any dependent app(s). The DataFutures thus produced by an app are accessible through the ``outputs`` attribute of the AppFuture.
-DataFutures inherit from and extend Python's `concurrent library <https://docs.python.org/3/library/concurrent.futures.html>`_.
+
+When calling an app that produces files as outputs, Parsl requires that a list of output files be specified (as a list of `File` objects pass in via the ``outputs`` keyword argument). A DataFuture for each output file is returned in the AppFuture when the app is executed, in the `outputs` attribute.
```
Comment (danielskatz) [inline docs/userguide/futures.rst:111]: pass in should be passed in

**Comment by danielskatz** [comment]
@benclifford  - I've made a bunch of small comments in-line.  In addition, I think it would make sense to add a little general text under the staging providers header that explains at that point the difference between the FTP, HTTP, HTTPS: separate task staging and FTP, HTTP, HTTPS: in-task staging sections that follow, something like - if you have this situation, use this part below...

**Comment by danielskatz** [comment]
The rest of this looks good to me

**Comment by danielskatz** [comment]
I've gone ahead and made the changes that are small.
I think the one more thing that is needed is to add a little general text under the staging providers header that explains at that point the difference between the FTP, HTTP, HTTPS: separate task staging and FTP, HTTP, HTTPS: in-task staging sections that follow, something like - if you have this situation, use this part below...

**Comment by yadudoc** [comment]
@danielskatz why don't we merge this now. Since you've already done a fair bit of testing with staging, can you try re-reading some of this and recommend changes in a separate PR?

**Comment by danielskatz** [comment]
I suggest we just merge this PR for now - It already has a bunch of changes I added a few weeks ago, and I don't have a good suggestion for the part that I thought still needed something more.

**Pull request merged**